docs/0005-hashed-n-tuple-id-layout.md

```diff
@@ -0,0 +1,235 @@
+
+If `tupleSize` is set to `0`, then no tuples are created and `numberOfTuples` MUST also equal `0`.
+
+The product of `tupleSize` and `numberOfTuples` MUST be less than or equal to the number of characters in the hex encoded digest.
+
+#### numberOfTuples
+
+`numberOfTuples` determines how many tuples to create from the digest. The tuples are used as directory names, and each successive directory is nested within the previous. The default value is `3`, which means that every OCFL object root will be 4 directories removed from the OCFL storage root, 3 tuple directories plus 1 encapsulation directory. Increasing this value increases the depth of the OCFL storage hierarchy.
+
+If `numberOfTuples` is set to `0`, then no tuples are created and `tupleSize` MUST also equal `0`.
+
+The product of `numberOfTuples` and `tupleSize` MUST be less than or equal to the number of characters in the hex encoded digest.
+
+## Procedure
+
+The following is an outline of the steps to follow to map an OCFL object identifier to an OCFL object root path using this extension:
+
+1. The OCFL object identifier is encoded as UTF-8 and hashed using the specified `digestAlgorithm`.
+2. The digest is encoded as a lower-case hex string.
+3. Starting at the beginning of the digest and working forwards, the digest is divided into `numberOfTuples` tuples each containing `tupleSize` characters.
+4. The tuples are joined, in order, using the filesystem path separator.
+5. The OCFL object identifier is URL-encoded to create the encapsulation directory name: all characters that aren't letters, digits, "\_" or "-" are changed to the %xx equivalent of their UTF-8 bytes. For example, "..hor/rib:lè-$id" becomes "%2e%2ehor%2frib%3al%c3%a8-%24id" as the encapsulation directory name.
+6. The whole encapsulation directory name is made lower-case, and is joined on the end of the path.
+
+## Examples
+
+### Example 1
+
+This example demonstrates what the OCFL storage hierarchy looks like when using this extension's default configuration.
+
+#### Parameters
+
+It is not necessary to specify any parameters to use the default configuration. However, if you were to do so, it would look like the following:
+
+```json
+{
+    "digestAlgorithm": "sha256",
+    "tupleSize": 3,
+    "numberOfTuples": 3
+}
+```
+
+#### Mappings
+
+| Object ID | Digest | Object Root Path |
+| --- | --- | --- |
+| object-01 | 3c0ff4240c1e116dba14c7627f2319b58aa3d77606d0d90dfc6161608ac987d4 | `3c0/ff4/240/object-01` |
+| ..hor/rib:le-$id | 487326d8c2a3c0b885e23da1469b4d6671fd4e76978924b4443e9e3c316cda6d | `487/326/d8c/%2e%2ehor%2frib%3ale-%24id` |
+
+#### Storage Hierarchy
+
+```
+[storage_root]/
+├── 0=ocfl_1.0
+├── ocfl_layout.json
+├── extensions/
+│   └── 0005-hashed-n-tuple-id-layout/
+│       └── 0005-hashed-n-tuple-id-layout.json
```

Comment:
I need to fix this in my PR too, but just noting that, based on this PR https://github.com/OCFL/extensions/pull/13, it sounds like the intent is for storage layout json to be written to the storage root and not within an extension directory.

Reply:
fixed